tools/lambda-promtail/lambda-promtail/main.go

```diff
@@ -25,19 +26,22 @@ const (
 
 	maxErrMsgLen = 1024
 
-	invalidExtraLabelsError = "invalid value for environment variable EXTRA_LABELS. Expected a comma separated list with an even number of entries. "
+	invalidExtraLabelsError  = "invalid value for environment variable EXTRA_LABELS. Expected a comma separated list with an even number of entries. "
+	invalidExtraHeadersError = "invalid value for environment variable EXTRA_HTTP_HEADERS. Expected a comma separated list with an even number of entries."
 )
 
 var (
-	writeAddress                                                             *url.URL
-	username, password, extraLabelsRaw, dropLabelsRaw, tenantID, bearerToken string
-	keepStream                                                               bool
-	batchSize                                                                int
-	s3Clients                                                                map[string]*s3.Client
-	extraLabels                                                              model.LabelSet
-	dropLabels                                                               []model.LabelName
-	skipTlsVerify                                                            bool
-	printLogLine                                                             bool
+	writeAddress                                                                              *url.URL
+	username, password, extraLabelsRaw, dropLabelsRaw, tenantID, bearerToken, extraHeadersRaw string
+	keepStream                                                                                bool
+	batchSize                                                                                 int
+	s3Clients                                                                                 map[string]*s3.Client
+	extraLabels                                                                               model.LabelSet
+	dropLabels                                                                                []model.LabelName
+	skipTlsVerify                                                                             bool
+	printLogLine                                                                              bool
+	extraHeaders                                                                              map[string]string
+	httpHeaderKeyRegex                                                                        = regexp.MustCompile("([A-Za-z0-9-])+")
```

Comment:
I think the set of characters that are valid for HTTP header keys is much more than this, is it not?

Reply:
You seem to be right about this from an RFC perspective, since AFAICS there is no real specification for the format of a valid HTTP header field name. Nevertheless some services still impose such restrictions - especially the ones stated by AWS ELB led me to idea to implement this:

- https://docs.aws.amazon.com/elasticloadbalancing/latest/application/application-load-balancers.html#:~:text=default%20is%20defensive.-,routing.http.drop_invalid_header_fields.enabled,-Indicates%20whether%20HTTP
- https://developers.cloudflare.com/rules/transform/request-header-modification/reference/header-format/
- https://pkg.go.dev/golang.org/x/net/http/httpguts#ValidHeaderFieldName